pbiapi/pbiapi.py

```diff
@@ -1,400 +1,327 @@
+            "client_secret": self.client_secret,
         }
-        headers = {
-            'Content-Type': "application/x-www-form-urlencoded",
-        }
-
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
         response = requests.post(self.url, data=payload, headers=headers)
 
-        if response.status_code == 200:
-            self.token = response.json()['access_token']
-            self.tokenExpiration = datetime.datetime.utcnow() + datetime.timedelta(seconds=3600)  
-
-            self.headers = {
-                'Content-Type': "application/x-www-form-urlencoded",
-                'Authorization': "Bearer " + self.token
-            }           
-            return True
+        if response.status_code == HTTP_OK_CODE:
+            self.token = response.json()["access_token"]
+            self.token_expiration = datetime.datetime.utcnow() + datetime.timedelta(hours=1)
+            self.headers = {**headers, **self.get_auth_header()}
         else:
-            print(response.status_code)
-            print(response.text)
-            return False
+            self.force_raise_http_error(response)
+
+    @property
+    def workspaces(self):
+        return self._workspaces or self.get_workspaces()
 
-   
-    #-----WORKSPACE FUNCTIONS----- 
-    @checkToken 
-    def getWorkspaces(self):        
-        url = "https://api.powerbi.com/v1.0/myorg/groups" 
+    @check_token
+    def get_workspaces(self):
+        url = self.base_url + "groups"
         response = requests.get(url, headers=self.headers)
-        if response.status_code == 200:
-            self.workspaces = response.json()['value']
-            return True
-        else:
-            return False
-     
-    def findWorkspaceIdByName(self,name):
-        self.getWorkspaces()
-        if self.workspaces != None:
-            return next((item['id'] for item in self.workspaces if item["name"] == name),None)
+
+        if response.status_code == HTTP_OK_CODE:
+            self._workspaces = response.json()["value"]
+            return self._workspaces
         else:
-            return None
-    
-    @checkToken
-    def createWorkspace(self,name):
-        #Check if workspace exists already
-        url = "https://api.powerbi.com/v1.0/myorg/groups?$filter=contains(name,'{name}')".format(name = name)
+            print("Failed to fetch workspaces!")
+            self.force_raise_http_error(response)
+
+    def find_workspace_id_by_name(self, name: str, raise_if_missing: bool = False):
+        for item in self.workspaces:
+            if item["name"] == name:
+                return item["id"]
+        if raise_if_missing:
+            raise RuntimeError(f"No workspace was found with the name: '{name}'")
+
+    @check_token
+    def create_workspace(self, name):
+        # Check if workspace exists already:
+        url = self.base_url + f"groups?$filter=contains(name,'{name}')"
```

Comment:
Should contains be equals?

Reply:
Yes it could. I just copied the code example to get it working but I believe this an odata filter so you could use:
`$filter=name eq '{name}'`

Reply:
Adressed in https://github.com/scottmelhop/PowerBI-API-Python/pull/2/files